Okta spring security feature flags update (#704)
oktadeveloper/okta-blog: _source/_posts/2020-10-09-okta-spring-security-feature-flags.md

> If you're interested in learning more about OIDC and OAuth 2.0, [here](https://developer.okta.com/blog/2017/07/25/oidc-primer-part-1) and [here](https://oauth.com) are good places to start. Look for more links to posts on OIDC and OAuth 2.0 at the end of this post.

### Register for an Okta Org and Create an OIDC App

{% include setup/cli.md type="web" framework="Okta Spring Boot Starter"
  loginRedirectUri="http://localhost:8080/login/oauth2/code/okta"

>   logoutRedirectUri="http://localhost:8080" %}

Reviewer: These redirects are the defaults for the Okta Spring Boot starter, so you can remove them.

#### Add Users to your Okta Org

> Head on over to http://{yourOktaDomain} and login.

Reviewer: ```suggestion
Run `okta login` and open the resulting URL in your browser. Sign in to the Okta Admin Console.
```

> Click **Directory > People** on the menu on the left. Here, you'll see the user already created for you with the name and email you submitted to create the Okta org.

Reviewer: ```suggestion
Then, go to **Directory** > **People**. Here, you'll see the user already created for you with the name and email you submitted to create the Okta org.
```

#### Add Groups to your Okta Org

> Choose **Directory > Groups** from the top level menu. Here, you'll see the built in `Everyone` group. As you might imagine, every current and new user is automatically added to this group.

Reviewer: ```suggestion
Navigate to **Directory** > **Groups**. Here, you'll see the built-in `Everyone` group. As you might imagine, every current and new user is automatically added to this group.
```

Click **Add Group**. This will bring you to the input form for adding a new group:

> {% img blog/okta-split-spring-security/new-group.png alt:"New User" width:"600" %}

Reviewer: Remove this line and delete the image. We don't want any images of the old developer console in blog posts.

Reviewer: Sorry, I didn't realize you created a new screenshot. Ignore my previous comment. ;)

Enter **BETA_TESTER** in both the **Name** and **Group Description** fields. Click **Add Group**.

Click the link to the newly created **BETA_TESTER** group. Click **Manage People**. Click on each of **Tina**, **Gene**, and **Louise** on the left. Notice that those users are moved from the left to the right, indicating that they will be members of the group. Click **Save** to finalize these changes.

Now, you're Okta org is configured with five users (the Belcher family), 3 of whom belong to the **BETA_TESTER** group.

### Configure the OIDC App to Return Groups Claim

The last Okta configuration step is to make sure that the list of groups a user belongs to is returned when a user authenticates. You'll see below that this integrates very easily with Spring Security.

> Click **Security > API** from the top menu bar. Click the **default** link under the list of **Authorization Servers** (it should be the only one right now).

Reviewer: ```suggestion
Go to **Security** > **API** and select the **default** authorization server (it should be the only one right now).
```

{% img blog/okta-split-spring-security/spring-initializr.png alt:"Initializr" width:"600" %}

Spring Initializr even makes it easy to load a pre-configured project from a [direct link](https://start.spring.io/#!type=maven-project&language=java&platformVersion=2.3.9.RELEASE&packaging=jar&jvmVersion=11&groupId=com.okta.examples&artifactId=okta-split-example&name=okta-split-example&description=Feature%20Flags%20with%20Okta%2C%20Split%20and%20Spring%20Security&packageName=com.okta.examples.okta_split_example&dependencies=web,thymeleaf,okta).

And, you can download the project from the command line with:

```bash
curl -G \
--data 'type=maven-project' \
--data 'language=java' \

> --data 'bootVersion=2.3.9.RELEASE' \

Reviewer: If we're going to update the Spring Boot version here, we should update it in the example app on GitHub too. Then, add a changelog that links to both. For the most part, I haven't been updating versions unless it's Spring Boot 2.1 (because https://github.com/oktadeveloper/okta-blog/issues/651).

Author: I changed it because the direct link in the post results in this message in spring initalizr web:
The following attributes could not be handled:
Spring Boot 2.3.4.RELEASE is not available, 2.3.10 has been selected.

How should I proceed?

Author: Now the same happens with 2.3.9!
The following attributes could not be handled:
Spring Boot 2.3.9.RELEASE is not available, 2.3.10 has been selected.

Author: Do I revert to 2.3.4 then?

Reviewer: Yes, I would. As long as start.spring.io changes to a 2.3.x version, I think it's OK.

Author: Done

### Integrate with Okta

> In order for this Spring Boot app to connect to your Okta org you need to rename `src/main/resources/application.properties` to `src/main/resources/application.yml`, and just set the three configuration parameters you got before when creating the client app with Okta CLI:

Reviewer: ```suggestion
In order for this Spring Boot app to connect to your Okta org you need to rename `src/main/resources/application.properties` to `src/main/resources/application.yml`, and just set the three configuration parameters you got before when creating the client app with the Okta CLI:
```

### Integrate with Okta

> In order for this Spring Boot app to connect to your Okta org you need to rename `src/main/resources/application.properties` to `src/main/resources/application.yml`, and just set the three configuration parameters you got before, when creating the client app with Okta CLI:

Reviewer: ```suggestion
In order for this Spring Boot app to connect to your Okta org you need to rename `src/main/resources/application.properties` to `src/main/resources/application.yml`, and just set the three configuration parameters you got before when creating the client app with Okta CLI:
```
